sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java

```diff
@@ -272,52 +339,210 @@ public Read withUpToDateThreshold(Duration upToDateThreshold) {
+      return builder().setPartitionKey(partitionKey).build();
+    }
+
+    public Write withPartitioner(KinesisPartitioner partitioner) {
+      return builder().setPartitioner(partitioner).build();
+    }
+
+    public Write withProducerProperties(Properties properties) {
+      return builder().setProducerProperties(properties).build();
+    }
+
+    public Write withAWSClientsProvider(AWSClientsProvider awsClientsProvider) {
+      return builder().setAWSClientsProvider(awsClientsProvider).build();
+    }
+
+    public Write withAWSClientsProvider(String awsAccessKey, String awsSecretKey, Regions region) {
+      return withAWSClientsProvider(awsAccessKey, awsSecretKey, region, null);
+    }
+
+    public Write withAWSClientsProvider(
+        String awsAccessKey, String awsSecretKey, Regions region, String serviceEndpoint) {
+      return withAWSClientsProvider(
+          new BasicKinesisProvider(awsAccessKey, awsSecretKey, region, serviceEndpoint));
+    }
+
+    public Write withRetries(int retries) {
+      return builder().setRetries(retries).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<byte[]> input) {
+      checkArgument(getStreamName() != null, "withStreamName() is required");
+      checkArgument(
+          (getPartitionKey() != null) || (getPartitioner() != null),
+          "withPartitionKey() or withPartitioner() is required");
+      checkArgument(
+          getPartitionKey() == null || (getPartitioner() == null),
+          "only one of either withPartitionKey() or withPartitioner() is possible");
+      checkArgument(getAWSClientsProvider() != null, "withAWSClientsProvider() is required");
+      checkArgument(streamExists(getAWSClientsProvider().getKinesisClient(), getStreamName()),
+          "Stream %s does not exist", getStreamName());
+      input.apply(ParDo.of(new KinesisWriterFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class KinesisWriterFn extends DoFn<byte[], Void> {
+      private final KinesisIO.Write spec;
+      private transient IKinesisProducer producer;
+      private transient KinesisPartitioner partitioner;
+
+      public KinesisWriterFn(KinesisIO.Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        // Init producer config
+        Properties props = spec.getProducerProperties();
+        if (props == null) {
+          props = new Properties();
+        }
+        KinesisProducerConfiguration config = KinesisProducerConfiguration.fromProperties(props);
+        // Fix to avoid the following message "WARNING: Exception during updateCredentials" during
+        // producer.destroy() call. More details can be found in this thread:
+        // https://github.com/awslabs/amazon-kinesis-producer/issues/10
+        config.setCredentialsRefreshDelay(100);
+
+        // Init Kinesis producer
+        producer = spec.getAWSClientsProvider().createKinesisProducer(config);
```

Comment:
Is it OK to create the producer in `Setup` instead of `StartBundle` ? For instance, in streaming mode, the producer can be open for a very long time. I wonder if Kinesis peer can "close" a producer open for a long time.
Thoughts ?

Reply:
Good point, @jbonofre. Actually, `KinesisProducer` (KP) is just a wrapper around native child process which is responsible for all communications with Kinesis endpoint. It's expected that only one instance of KP should be used per worker, creating new instance of KP is quite heavy operation (since it creates new system process) and all retries and error handling is incapsulated inside KP. I think, it would make sense to run long IT to see how it works under this condition.

Reply:
Agree as it's a heavy operation. I would like to check that we don't have any timeout or kick out from Kinesis with a long pipeline execution (like in streaming mode for instance). Thanks !

Reply:
@jbonofre I performed a simple test to test this behaviour:
1) I created a pipeline that reads data from Kafka and writes into Kinesis 
2) I run this pipeline with direct runner and sent new message to Kafka.
3) I checked that new records was stored into Kinesis stream
4) Then I was waiting for 2 hours and send another message to Kafka.
5) After I checked that another record was stored in Kinesis too.

So, for now, I think we can conclude that it's not an issue. Are you agree?

Reply:
It sounds good to me. Thanks !